File: app/src/main/java/no/nav/tilbakemeldingsmottak/consumer/saf/graphql/JournalpostToValidator.java
package no.nav.tilbakemeldingsmottak.consumer.saf.graphql;

import no.nav.tilbakemeldingsmottak.consumer.saf.journalpost.SafJournalpostTo;

import java.util.List;

import static no.nav.tilbakemeldingsmottak.consumer.saf.util.ValidationUtil.*;

public class JournalpostToValidator {

    public SafJournalpostTo validateAndReturn(SafJournalpostTo safJournalpostTo) {
        assertJournalpostFieldNotNull(SafJournalpostTo.DokumentInfo.class, safJournalpostTo.getDokumenter());
        validateDokumenter(safJournalpostTo.getDokumenter());

        return safJournalpostTo;
    }

    private void validateDokumenter(List<SafJournalpostTo.DokumentInfo> dokumenter) {
        dokumenter.forEach(this::validateDokument);
    }

    private void validateDokument(SafJournalpostTo.DokumentInfo dokumentInfo) {
        assertDokumentFieldNotNullOrEmpty("dokumentInfoId", dokumentInfo.getDokumentInfoId());
        validateDokumentVarianter(dokumentInfo.getDokumentvarianter());
    }

    private void validateDokumentVarianter(List<SafJournalpostTo.Dokumentvariant> dokumentvarianter) {
        dokumentvarianter.forEach(this::validateAndReturnDokumentVariant);
    }

    private void validateAndReturnDokumentVariant(SafJournalpostTo.Dokumentvariant dokumentvariant) {
        assertNotNullOrEmpty("variantformat", dokumentvariant.getVariantformat());
    }
}

File: app/src/main/kotlin/no/nav/tilbakemeldingsmottak/consumer/saf/graphql/JournalpostToValidator.kt
package no.nav.tilbakemeldingsmottak.consumer.saf.graphql

import no.nav.tilbakemeldingsmottak.saf.generated.hentjournalpost.Journalpost
import no.nav.tilbakemeldingsmottak.consumer.saf.util.ValidationUtil
import no.nav.tilbakemeldingsmottak.saf.generated.hentjournalpost.DokumentInfo
import no.nav.tilbakemeldingsmottak.saf.generated.hentjournalpost.Dokumentvariant
import java.util.function.Consumer

class JournalpostToValidator {
    fun validateAndReturn(safJournalpostTo: Journalpost): Journalpost {

        ValidationUtil.assertJournalpostFieldNotNull(
            DokumentInfo::class.java,
            safJournalpostTo.dokumenter
        )
        validateDokumenter(safJournalpostTo.dokumenter!!)
        return safJournalpostTo
    }

    private fun validateDokumenter(dokumenter: List<DokumentInfo?>) {
        dokumenter.forEach(Consumer { dokumentInfo -> validateDokument(dokumentInfo) })
    }

    private fun validateDokument(dokumentInfo: DokumentInfo?) {
        ValidationUtil.assertJournalpostFieldNotNull(
            DokumentInfo::class.java,
            dokumentInfo
        )
        ValidationUtil.assertDokumentFieldNotNullOrEmpty("dokumentInfoId", dokumentInfo?.dokumentInfoId)
        validateDokumentVarianter(dokumentInfo?.dokumentvarianter)
    }

    private fun validateDokumentVarianter(dokumentvarianter: List<Dokumentvariant?>?) {
        ValidationUtil.assertJournalpostFieldNotNull(
            Dokumentvariant::class.java,
            dokumentvarianter
        )
        dokumentvarianter?.forEach(Consumer { dokumentvariant ->
            validateAndReturnDokumentVariant(
                dokumentvariant
            )
        })
    }

    private fun validateAndReturnDokumentVariant(dokumentvariant: Dokumentvariant?) {
        ValidationUtil.assertNotNullOrEmpty("variantformat", dokumentvariant?.variantformat?.name)
    }
}
